x-pack/platform/plugins/shared/onechat/server/routes/esql_tools.ts

```diff
@@ -0,0 +1,373 @@
+          return response.notFound();
+        }
+
+        try {
+          const { esql: esqlToolClientService } = getInternalServices();
+          const client = await esqlToolClientService.getScopedClient({ request });
+
+          const tools = await client.list();
+
+          return response.ok({
+            body: {
+              tools,
+            },
+          });
+        } catch (error) {
+          return response.customError({
+            statusCode: 500,
+            body: {
+              message: `Error: ${error.message}`,
+            },
+          });
+        }
+      })
+    );
+
+  router.versioned
+    .post({
+      path: '/api/chat/tools/esql',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
+      },
+      access: 'public',
+      summary: 'ESQL Tool API',
+      description: TECHNICAL_PREVIEW_WARNING,
+      options: {
+        tags: ['esql'],
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: {
+            body: schema.object({
+              id: schema.maybe(schema.string()),
+              name: schema.string(),
+              description: schema.string(),
+              query: schema.string(),
+              params: schema.recordOf(
+                schema.string(),
+                schema.object({
+                  type: schema.string(),
+                  description: schema.string(),
+                })
+              ),
+              meta: schema.object({
+                tags: schema.arrayOf(schema.string()),
+              }),
+            }),
+          },
+        },
+      },
+      wrapHandler(async (ctx, request, response) => {
+        const { uiSettings } = await ctx.core;
+        const enabled = await uiSettings.client.get(ESQL_TOOL_API_UI_SETTING_ID);
+
+        if (!enabled) {
+          return response.notFound();
+        }
+        try {
+          const { esql: esqlToolService } = getInternalServices();
+          const client = await esqlToolService.getScopedClient({ request });
+          const toolid = request.body.id || uuidv4();
+          const now = new Date();
+
+          const tool: EsqlToolCreateRequest = {
+            id: toolid,
+            name: request.body.name,
+            description: request.body.description,
+            query: request.body.query,
+            params: request.body.params || {},
+            meta: {
+              tags: request.body.meta.tags || [],
+              providerId: esqlToolProviderId,
+            },
+            created_at: now.toISOString(),
+            updated_at: now.toISOString(),
+          };
+
+          const createResponse = await client.create(tool);
+
+          return response.ok({
+            body: {
+              tool: createResponse,
+            },
+          });
+        } catch (error) {
+          return response.customError({
```

Comment:
why `customError`?  pretty sure `response` object has sth like `.internalServerError` that wraps the 500 response 

use `customError` only when there is no built in method in the responsefactory